### DISCUSSION TITLE
Could someone please explain the nuance of mesh levelling to me

### DISCUSSION THREAD

**Discussion by vaderag** [Jyers/Marlin#212] [Q&A] [answered]
First of all, I absolutely LOVE this firmware - thank you so much.
But I'm having some issues with levelling...

I've enabled the mesh levelling firmware and it seems to work great _initially_ 

I do the following:
- Manual level
- Mesh level via home screen, move up and down as necessary in each spot with my post it note
- Save configuration

I then run a bed level test print, all seems good, offset on the fly as needed during this and take the offset I needed and dump it in my slicer (say -0.05)

I then print something, generally all good, but might find that things are still a little high, so i tweak the z-offset on the fly by (say) -0.02 to give a bit of squish. Adjust the offset in slicer accordingly to (say) -0.07

But then next print (or the print after... at _some point_ it'll be off - and usually by either the big or the little number above, and I'll need to drop it by 0.07 again. Basically like it's completely forgetting mesh at the start of that print, or that it's not set in the slicer.

I haven't yet managed to work out what causes it, but I have noticed that occasionally the z-offset will stay plugged in between prints and sometimes it wont. 

So, i'm just wondering if I'm doing something wrong, or if there is something I'm getting confused in the move to mesh levelling and slicer settings etc

Thanks!

**Comment by xantrk** [comment]
My experience is similar, unfortunately. It diverges from level to bad faster than conventional manual leveling for me. I am hopeful about UBL, currently testing.

**Reply by vaderag** [reply]
I'm really glad it's not just me! TBH, I was starting to think I'd gone mad!! 
What is UBL? Mesh would be perfect for me if it were stable

**Reply by xantrk** [reply]
> I'm really glad it's not just me! TBH, I was starting to think I'd gone mad!!
> What is UBL? Mesh would be perfect for me if it were stable

More advanced mesh leveling basically. It released today for this firmware!

**Comment by wbvczar** [comment]
Make sure to include the command M420 S1 in the “Start G-code” in your slicer settings. This will load the mesh that you created.

**Reply by vaderag** [reply]
Oh man. Now I feel stupid... I didn't realise I needed to add start code to use it, just assumed it would (since it was in the firmware!)

**Reply by danrough** [reply]
Well, I just came here to let you know that you're not alone - I've been printing test prints all weekend and wondering why the mesh levelling that I had conducted had seemingly had no effect. And now I too realise that I need to add some start code.

**Reply by wbvczar** [reply]
I was just looking at the latest code from jyers and I see this in the code (UBL-NoProbe-3x3/Configuration.h).

/**
 * Normally G28 leaves leveling disabled on completion. Enable one of
 * these options to restore the prior leveling state or to always enable
 * leveling immediately after G28.
 */
#define RESTORE_LEVELING_AFTER_G28
//#define ENABLE_LEVELING_AFTER_G28

Can someone verify if this does the same as "M420 S1" making it not needed in the "Start G-code"?

**Reply by Jyers** [reply]
> Can someone verify if this does the same as "M420 S1" making it not needed in the "Start G-code"?

This is not necessarily the case, this setting will restore the previous mesh state after a home. This means if the mesh was disabled before homing, as would be the case after the printer is turned off and on again, the mesh will still be disabled after it is homed. It is always good to have M420 S1 in your start Gcode to ensure that the mesh is enabled for each print.

**Reply by vaderag** [reply]
I think that was the mistake I made that I saw that bit saying that it re-enabled levelling and assumed nothing else was needed!
Will run some tests but suspect that may well be the issue!

**Reply by vaderag** [reply]
Actually, going to go out on a limb and say that this still isn't the case.
Added the M420 S1 to my start code, just started a print with something I'd sliced at -0.08 (actually forgetting), expecting to have to run in and adjust up to save my bed, but actually still needed lowering again...
very confused

**Reply by danrough** [reply]
@vaderag I haven't had a chance to try this, yet. I was going to experiment with these changes tonight. If you have the time could you note the steps you followed - I can then follow them too, to see if I get a similar outcome. That might be helpful if this becomes an issue and we want to report it as such. We'll have some data that we can provide :)

**Reply by vaderag** [reply]
@danrough the steps I've followed are broadly in the initial post.

I've just come to my printer this morning, with the M420 in my gcode and it's got even worse! Had to drop by -0.12 on the fly to get to the level this morning

It feels like it's overcompensating or something on save, but I'm struggling to fully identify.
I'm going to re-level later and will start noting from the initial mesh level exactly what occurs - would be great if you could do the same!

**Reply by danrough** [reply]
Okay, I'll give this a go later on today. Did you see this comment below https://github.com/Jyers/Marlin/discussions/212#discussioncomment-461256 I wonder if that might be the solution to your problem?

**Reply by danrough** [reply]
I didn't manage to get to this this evening, unfortunately. I'll try to find some time at some point tomorrow. It might end up being something I do over the weekend though.

**Reply by Turismo05** [reply]
I was thinking the same thing, I noticed I was getting a much better first layer after using mesh levelling I didn't know about adding extra g-code in my slicer.

**Comment by xantrk** [comment]
I didn't want to misguide but I already added the gcode, and still it
somehow goes off level more often than regular manual level. Didn't had
enough chance to test UBL yet. It is as if I need to offset again whenever
it doesn't work.

**Reply by vaderag** [reply]
I just posted the same actually above - "Actually, going to go out on a limb and say that this still isn't the case.
Added the M420 S1 to my start code, just started a print with something I'd sliced at -0.08 (actually forgetting), expecting to have to run in and adjust up to save my bed, but actually still needed lowering again...
very confused"

**Comment by rorywquin** [comment]
I use M420 S1 Z2 and it works for me. 

I believe 2 = Height in MM not layers.

From Marlin:

Set Z fade height (Requires ENABLE_LEVELING_FADE_HEIGHT)

With Fade enabled, bed leveling correction is gradually reduced as the nozzle gets closer to the Fade height. Above the Fade height no bed leveling compensation is applied at all, so movement is machine true.
Set to 0 to disable fade, and leveling compensation will be fully applied to all layers of the print.

**Comment by vaderag** [comment]
Okay, so I'm documenting everything I'm doing since I've _just_ relevelled the bed.

### **a. Manual Levelling Process**:
- Turn on Machine
- Preheat -> Warmup (190 / 60)
- Prepare -> Manual Levelling
- Checked Machine Z-Offset is 0
- 4 corners & center level manually with paper
- Back, Back, Control -> **Store Settings** 
- "Settings Stored" confirmed on screen

### **b. Mesh Levelling Process**:
- (Immediately after above)
- Level on Home Screen
- Run through the 9 points 
- Z-offset down (from the default 0.2 it gives) at each point
- Next
- Save mesh at the end
- Control -> **Store Settings**
- "Settings Stored" confirmed on screen

### **1. Print One**:
- Slice in Cura with Z-Offset set to 0
- Upload to OctoPi
- Print
- When doing the skirt, adjust z-offset down as needed using up and down (ended up at -0.07 which is what I would expect as my post it is -0.07 ish and that number worked on non-mesh levelling before)
- Let the print finish, all good.

### **POST PRINT One**:
- Update Cura to have z-offset as -0.07 per above so manual compensation should not be needed in second print


Watch this space...

**Reply by vaderag** [reply]
> Okay, so I'm documenting everything I'm doing since I've _just_ relevelled the bed.
> 
> ### **a. Manual Levelling Process**:
> * Turn on Machine
> * Preheat -> Warmup (190 / 60)
> * Prepare -> Manual Levelling
> * Checked Machine Z-Offset is 0
> * 4 corners & center level manually with paper
> * Back, Back, Control -> **Store Settings**
> * "Settings Stored" confirmed on screen
> 
> ### **b. Mesh Levelling Process**:
> * (Immediately after above)
> * Level on Home Screen
> * Run through the 9 points
> * Z-offset down (from the default 0.2 it gives) at each point
> * Next
> * Save mesh at the end
> * Control -> **Store Settings**
> * "Settings Stored" confirmed on screen
> 
> ### **1. Print One**:
> * Slice in Cura with Z-Offset set to 0
> * Upload to OctoPi
> * Print
> * When doing the skirt, adjust z-offset down as needed using up and down (ended up at -0.07 which is what I would expect as my post it is -0.07 ish and that number worked on non-mesh levelling before)
> * Let the print finish, all good.
> 
> ### **POST PRINT One**:
> * Update Cura to have z-offset as -0.07 per above so manual compensation should not be needed in second print
> 
> Watch this space...


### **2. Print Two**:
* Cura now has offset -0.07 so in theory, no more adjusting needed.
* Start print. 
* Minor adjustment needed needed +0.01 as was a little close. I'm gonna say that's near as none tho 🤷  Maybe this was all in my head... will continue to log prints for a bit

**Reply by Schnabulation** [reply]
Can I suggest something? Why do you feel the need to set the Z-offset in Cura and not directly on the printer (and then stored to EEPROM)?

When you set it in Cura it is baked into your G-code - however when you set it on the printer you can adjust it all the time. You just need to remember to store it in EEPROM.

**Reply by vaderag** [reply]
> Can I suggest something? Why do you feel the need to set the Z-offset in Cura and not directly on the printer (and then stored to EEPROM)?
> 
> When you set it in Cura it is baked into your G-code - however when you set it on the printer you can adjust it all the time. You just need to remember to store it in EEPROM.

I'm not quite sure I knew I could store the z-offset in my EEPROM. Obviously there are settings in the mesh, and during the print, but generally post print it drops the z-offset (or at least post re-boot). If I save the config following a print with adjustments will it save the offset?

**Reply by Schnabulation** [reply]
Yes, it should do so. Basically when you change any setting - be this via the display or via a G-code command it will store it in the RAM. When you restart the printer it will be gone and only the settings in EEPROM will be loaded. But, when you go to "Control" -> "Store settings" it will store all settings from RAM to EEPROM, so they stick after a reboot.

Do you have a device connected via USB with access to the serial console of the printer? You can send the command `M503` to display the current settings in RAM. Then you could restart the printer and send M503 again - now you should see all settings in EEPROM.

**Reply by Schnabulation** [reply]
...also, when you go to "Control" -> "Reset to default" the EEPROM will be cleared and only the settings from the firmware config-files will be loaded.

So basically you have 3 places to store settings:

- in configuration.h and configuration_adv.h when compiling the firmware
- in EEPROM
- in RAM (or SRAM as it's actually called)

**Reply by vaderag** [reply]
Thanks for the detail @Schnabulation 
Presumably an M503 in Octoprint will give me that so will take a look.

Out of interest, related, if I update the firmware (say to the latest Jyers), is there an easy way to backup/re-import the settings i changes (e.g. the mesh) or do I need to start from scratch?

**Reply by Schnabulation** [reply]
It is highly advisable to reset the EEPROM after a firmware change or upgrade. I know it's a PITA to redo the mesh but it really is recommended to start from scratch. I had a quick google and it looks like you can export the mesh values via G-codes and then re-import them. But I have never done this in the past.

**Reply by Jyers** [reply]
Just a note that with UBL, you do not have to redo the mesh. Even after resetting eeprom or flashing new firmware, you can simply go to the leveling menu and select "load mesh" and it will load the last mesh from before the reset.

**Reply by Schnabulation** [reply]
Srly? Is it stored in another part of the EEPROM that does not get cleared when told to do so?

**Reply by Jyers** [reply]
Yeah, I don't know the technical details, but there must be some section of eeprom that is not cleared

**Reply by Jyers** [reply]
This is from the docs: "We know it takes a lot of effort to create good working Mesh data, so UBL stores its mesh data in a separate location where it won’t be affected by EEPROM errors or version changes."

**Reply by vaderag** [reply]
> Just a note that with UBL, you do not have to redo the mesh. Even after resetting eeprom or flashing new firmware, you can simply go to the leveling menu and select "load mesh" and it will load the last mesh from before the reset.

Oh wow, that's super handy. I guess I should look at using UBL then rathe than Mesh 
Any cons to using UBL instead that you're aware of?

**Reply by vaderag** [reply]
Also, can someone advise how to clear the EEPROM? I thought it used the SD card but googling I'm seeing that actually might use onboard now?

**Reply by Schnabulation** [reply]
The EEPROM was simulated on the SD card when using the crappy Creality stock firmware - but our board actually has onboard EEPROM storage, which is used by official Marlin (and Jyers fork for that matter).

You can clear it by either sending a M502 followed by a M500 via terminal or going to the "Control" menu, "Reset to defaults" then "Store settings". You need so send M500 or "Store settings" so the EEPROM gets overwritten with the factory defaults.

**Reply by Jyers** [reply]
> Any cons to using UBL instead that you're aware of?

The only real con is its a bit more complicated, but that is to be expected with the amount of features it adds.

**Reply by rorywquin** [reply]
I'm a bit confused
Step a.6 ...Back, Back, Control -> Store Settings 
What are you storing? if you've levelled with paper and the bed screws, it means nothing to the firmware.

**Comment by OB1-OB2** [comment]
This is what I and it works really well for me:

1. Use manual leveling to get the center and the corners have the exact same amount of friction when i put a piece of paper under the nozzle.
2. Use z-offset menu to lower the z-offsetto increase the friction to the point where the paper is hard to pull out, but it still pulls out (I do this at the center of the bed). Save the z-offset.
3. Run 5x5 ABL auto level thingy and save to eprom when prompted. Modify start gcode with M420 S1 after G28 to read the mesh.
4. do a print that prints a square to all 4 corners and center of bed, and adjust z-offset if the squares are not smooth (usually I lower the z-offset more), save z-offset. Repeat step 4 until i am happy with all 5 squares.

When step 4 results in all 5 squares being as smooth as a baby's butt, i am satisfied and move on to printing things.

If I'm doing something wrong, please let me know... I can't tell if the ABL mesh is actually doing much or not as my bed is pretty level on it's own.

**Reply by pesimistanblue** [reply]
Please try a powercycle (turn off and on your printer), and try to print something. Does your bed stay level after this?

**Reply by OB1-OB2** [reply]
Sorry about the delay. The bed does stay level after power cycle,

but like I said, My bed is pretty level and I’m not sure if the mesh is being applied or not

**Reply by Jyers** [reply]
A good way to check if the mesh is being applied is to print a test print that covers a large area of the bed and watch the z coupler to see if it is moving. It won't be much, but you should see it turn slightly during large moves.

**Reply by slingshot322** [reply]
@Jyers would you not see the Z value change on the LCD like you would with the X and Y axis as it prints if it's using the mesh level settings?

**Reply by Jyers** [reply]
No you won't. The mesh modifies the z position on a much lower level then the high level current_position used by the screen and most things. So you won't see any change on the screen or on any host applications. The best way is to just watch closely for z movement like I described.

**Comment by vaderag** [comment]
I'm getting to the point where i'm convinced that Mesh levelling is causing me issues
I'm getting extruder slipping at around the 2mm mark where the fade is supposed to finish and before that I can't get a clean layer out, as the nozzle just seems to coast through the print
I'm going to try turning it off when I have a bit more time  and see if reverting back to previous levelling methods is enough to solve

Presumably removed the M420 is sufficient to just use the manual levelling settings?

**Reply by Tuncay-Ayhan** [reply]
Yes. Because G28 (Autohome) disables it.

**Comment by vaderag** [comment]
So, just to come back to this, Mesh was DEFINITELY messing up my prints some how. 
I disabled two days ago and reverted to a non-mesh firmware and a bunch of issues (mostly extruder skipping at certain layers) I was having have just gone.

**Comment by tinooo** [comment]
Hi,
I can totally relate to this issue. My setup/results have been a little different, but with roughly the same experience. And I've also been at a point, not beeing sure if it was just my imagination... But I'm quite confident, that there is some issue somewhere.
Here the Thread:
https://github.com/Jyers/Marlin/discussions/173

I used a ABL probe, but roughly the same experience. Unfortunately, I used a capacitive one, which seems to be too inacurate over time, to ensure the homing point of next days print is identical to the homing point of todays print. So that had quite an impact on my z-offset, basically every few prints.

Now I've received the BL-touch a few days ago and installed it. Using a selfcompiled 5x5 version, of the templater from jyers fw. Now the rather huge Z-offset issues are gone (mostlikely due to the capacitive probe), but I still seem to have some issues, similar to yours.

I can't get it why. I've basically checked everything i can imagine (all screws, z-rod, bearings, ...). Also SW vise, I've put plenty of effort, to make sure not having some different values in eeeprom, FW, Ram, etc. just to make sure excluding this as an error point. Always put all numbers back in FW, recompiled, reset eeprom, restarted printer, etc.

But still it seems to be messing with me. I've no clue whats left to do here. Next step for me would be to try UBL.

I also encounter the issue, that when printing some calibration squres in each corner, I never get all of them perfect. There are always some a little to close/far away. Bed is prelevled, and Z-axis only doing realy tiny movements, alsmost not recognizable. I've also tried to have the bed Tilt by a few turns. ABL seems to do it's job. The compensation is now definitley noticable on z axis. But the results are as bad as before (actually even worse, from good to bad square, i'd assume at least 0.1mm difference).

**Reply by vaderag** [reply]
Odd thing on mine was calibration WAS perfect. It appeared to be some issue in the way fading is handled (at a best guess)
I literally have the same thing (some coasters) printed before and after with the only change being mesh on or off and without the print is clean, had no clicking, blemishes on the side disappeared.

**Comment by Jyers** [comment, marked as answer]
Gonna mark this as answered because its a little stale. If anything new comes up and we want to discuss more, maybe start a new discussion

**Reply by orangoo** [reply]
Hi all.

Just thought I'd add to this thread and say that I also had the same problem.

I was using v1.3.1. But I have upgraded to v1.3.2 and so far the manual mesh is being used. 
Thought I'd mention it as this could be the answer for many here.

thanks
